framework/src/modules/chain/bft/bft.js

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const assert = require('assert');
+const { HeadersList } = require('./headers_list');
+const blockHeaderSchema = require('./block_header_schema');
+const { validate } = require('../../../../src/controller/validator');
+
+/**
+ * Validate schema of block header
+ *
+ * @param {BlockHeader} blockHeader
+ * @return {boolean}
+ */
+const validateBlockHeader = blockHeader =>
+	validate(blockHeaderSchema, blockHeader);
+
+/**
+ * @typedef {Object} BlockHeader
+ * @property {string} blockId
+ * @property {int} height
+ * @property {int} maxHeightPreviouslyForged
+ * @property {int} prevotedConfirmedUptoHeight
+ * @property {int} activeSinceRound
+ * @property {string} delegatePublicKey
+ */
+
+class BFT {
+	constructor({ finalizedHeight, activeDelegates } = {}) {
+		assert(finalizedHeight !== undefined, 'Must provide finalizedHeight');
+		assert(activeDelegates !== undefined, 'Must provide activeDelegates');
+		assert(activeDelegates > 0, 'Must provide a positive activeDelegates');
+
+		// Set constants
+		this.ACTIVE_DELEGATES = activeDelegates;
+
+		// Threshold to consider a block prevoted
+		this.PRE_VOTE_THRESHOLD = Math.ceil((this.ACTIVE_DELEGATES * 2) / 3);
+
+		// Threshold to consider a block pre-committed (or finalized)
+		this.PRE_COMMIT_THRESHOLD = Math.ceil((this.ACTIVE_DELEGATES * 2) / 3);
+
+		// Limit to of blocks to must have to make any prevote-precommit verification on blocks (1 block less than 3 rounds)
+		this.PROCESSING_THRESHOLD = this.ACTIVE_DELEGATES * 3 - 1;
+
+		// Maximum headers to store (5 rounds)
+		this.MAX_HEADERS = this.ACTIVE_DELEGATES * 5;
+
+		this.headers = new HeadersList({ size: this.MAX_HEADERS });
+
+		// Height up to which blocks are finalized
+		this.finalizedHeight = finalizedHeight || 0;
+
+		// Height up to which blocks have pre-voted
+		this.prevotedConfirmedHeight = 0;
+
+		this.state = {};
+		this.preVotes = {};
+		this.preCommits = {};
+	}
+
+	/**
+	 * Add block header to BFT
+	 *
+	 * @param {BlockHeader} blockHeader
+	 * @return {BFT}
+	 */
+	addBlockHeader(blockHeader) {
+		// Validate the schema of the header
+		// To spy exported function in same module we have to call it as this
+		exportedInterface.validateBlockHeader(blockHeader);
```

Comment:
prefer to change the test methodology than affecting the source code.
In this case, you can spy on `validate`

Reply:
The right approach was using a `static` method while we already using class here. In initial review you tend to move it to top level function. There are known issues while testing the top level functions, the function reference inside the module would always be different than exported one. So if you spy on exported function, it does not work. @yatki also looked into this as well. That was one clean solution I find out to overcome this issue.